### PR TITLE
Automatic OTP2 smoke tests

### PR DESCRIPTION
**Pull request by leonardehrenfried**
This adds automatic smoke tests for the OTP2 branch. The purpose of those tests is a last sanity check that an upstream merge won't break the current IBI deployments.

A real Atlanta OTP graph is build from OSM and GTFS dependencies and an instance is started before HTTP requests are sent to it.

I've currently hosted two build input files on my own server:

- https://leonard.io/ibi/atlanta-2021-12-08.osm.pbf (OSM extract of the Atlanta area)
- https://leonard.io/ibi/cobblinc-flex-with-farezones.zip (Cobb flex GTFS)

Both of those should probably come from IBI-controlled infrastructure like an S3 bucket.

I've decided to run the tests with a separate workflow file as I think we only want to run it for PRs not pushes. It could however be integrated into the regular CI build. WDYT?

### CODE REVIEW

File: src/main/java/org/opentripplanner/routing/core/Money.java [hunk 1]
```diff
@@ -56,10 +56,6 @@ public void setCurrency(Currency currency) {
         this.currency = new WrappedCurrency(currency);
     }
     
-    public void setCurrency(WrappedCurrency currency) {
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/routing/core/Money.java]: This seems like a good change, but why was it made as part of this PR?

**Review by miles-grant-ibigroup — APPROVED** [review]
This looks great! Thanks a lot for separating them out from the main tests, I think it's a great idea to have them only run on PRs and to have them run by a separate command!

Using `make` means that using Windows will become even more difficult, but I think that's something we're ok with.

Also, I would add some more origin/destination pairs, but I think that can be done as a separate PR!

Otherwise, just one final small question about one change but for the most part this looks great thanks again!

**Comment by miles-grant-ibigroup** [comment]
Also maybe we should throw a blocked label on this until we get the files hosted on s3? What do you think @binh-dam-ibigroup

File: src/main/java/org/opentripplanner/routing/core/Money.java (hunk 1 above)
Comment (leonardehrenfried) [reply src/main/java/org/opentripplanner/routing/core/Money.java]: I removed this unused method as Jackson was getting confused as to which constructor should be used. In the end I couldn't use it anyway so will restore it to its previous state.

**Comment by leonardehrenfried** [comment]
So here comes the first time the smoke test catches something: it currently fails because in order to have somewhat predictable results I set the travel time to be on the next Monday, relative of the time of the test run.

The MARTA feed (https://itsmarta.com/google_transit_feed/google_transit.zip) however has only routes until December 17. This is in four days. Isn't this a little bit close?

**Comment by binh-dam-ibigroup** [comment]
> So here comes the first time the smoke test catches something: it currently fails because in order to have somewhat predictable results I set the travel time to be on the next Monday, relative of the time of the test run.
> 
> The MARTA feed (https://itsmarta.com/google_transit_feed/google_transit.zip) however has only routes until December 17. This is in four days. Isn't this a little bit close?

@leonardehrenfried Yes, a new MARTA feed will take effect on Dec. 18.

**Comment by leonardehrenfried** [comment]
So they publish the new feed after the old one has expired? :open_mouth:

**Comment by binh-dam-ibigroup** [comment]
> So they publish the new feed after the old one has expired? 😮

Usually they release it 1-2 days before.

**Comment by leonardehrenfried** [comment]
@miles-grant-ibigroup Do you have a preferred windows-compatible tool that I can use to script the download of the input files and the graph build?

**Comment by miles-grant-ibigroup** [comment]
The datatools e2e tests solve this by using our node-based otp-runner, but I don't think that's needed, especially since this is mostly designed to run in a ci!

**Comment by leonardehrenfried** [comment]
MARTA has now released the new feed whose routes start on Saturday means we _have_ to plan a route for next Monday for the test to work.

I've documented this in d541f799d1adf982a0f3d8aa2a138cfd056cccf1

File: smoke-tests/Readme.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+```
+# start in root OTP dir
+cd OpenTripPlanner
+mvn package -DskipTests
+cd smoke-tests
+make build-atlanta
+```
+
+After the graph is build you can then run OTP with:
```
Comment (binh-dam-ibigroup) [inline smoke-tests/Readme.md]: Typo - "built"

File: smoke-tests/Readme.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+## IBI Group OTP Smoke Tests
+
+This folder contains scripts and configuration files for running an OTP instance that can then be
+use to run smoke tests against. These tests are intended as a last sanity check when we merge
+upstream changes as these merges tend to be very large. It helps to reduce the burden on the
```
Comment (binh-dam-ibigroup) [inline smoke-tests/Readme.md]: "It helps ~to~ reduce"

File: smoke-tests/Readme.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+make build-atlanta
+```
+
+After the graph is build you can then run OTP with:
+
+```
+make run-atlanta
+```
+
+And finally run the smoke tests with
```
Comment (binh-dam-ibigroup) [inline smoke-tests/Readme.md:27]: Say it needs to run from the repo root folder.

File: src/test/java/org/opentripplanner/smoketest/SmokeTest.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is both a utility class and a category to select or deselect smoke tests during test
+ * execution.
+ * <p>
+ * By default, the smoke tests are not run when you execute `mvn test`.
+ * <p>
+ * If you want run them use the following command: `mvn test -P smoke-tests`
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/smoketest/SmokeTest.java]: "If you want to run them, use..."

File: src/test/java/org/opentripplanner/smoketest/SmokeTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        mapper.registerModule(new ParameterNamesModule(JsonCreator.Mode.PROPERTIES));
+        mapper.registerModule(module);
+    }
+
+    /**
+     * In order to have somewhat predictable results we get the route for the next Monday.
+     *
+     * FIXME: In the tech call on 2021-12-14 we said that we want to use the next business day, but
+     * as of 2021-12-15 this is not possible because the trips in the MARTA feed available to download
+     * start on 2021-12-18. After that date we can actually implement this logic.
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/smoketest/SmokeTest.java]: I think this part of the comment can be removed now. Or maybe change it to say we could expect failures in the days leading to a planned service change as the corresponding GTFS would not be available yet.

File: src/test/java/org/opentripplanner/smoketest/SmokeTest.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                .map(kv -> kv.getKey() + "=" + kv.getValue())
+                .collect(Collectors.joining("&"));
+
+        var uri = URI.create("http://localhost:8080/otp/routers/default/plan?" + urlParams);
+
+        return HttpRequest.newBuilder()
+                .uri(uri)
+                .GET()
+                .build();
+
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/smoketest/SmokeTest.java]: Remove this blank line.

File: src/test/java/org/opentripplanner/smoketest/SmokeTest.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    /**
+     * Sends an HTTP request to the OTP plan endpoint and deserializes the response.
+     */
+    static TripPlannerResponse sendPlanRequest(Map<String, String> params) {
+        var request = SmokeTest.buildPlanRequest(params);
+        LOG.info("Sending request to {}", request.uri());
+        TripPlannerResponse otpResponse;
+        try {
+
```
Comment (binh-dam-ibigroup) [inline src/test/java/org/opentripplanner/smoketest/SmokeTest.java]: Remove this blank line.

**Review by binh-dam-ibigroup — APPROVED** [review]
LGTM. Please update some comments where indicated before merging.

**Comment by leonardehrenfried** [comment]
Do you want to host the data input files on your infrastructure? It's fine for it to stay where they are if you don't.

**Comment by leonardehrenfried** [comment]
I've created a task for the input files to be moved to IBI infrastructure: https://teams.microsoft.com/l/entity/com.microsoft.teamspace.tab.planner/tt.c_19:2564d72f7a7e4ab2b45d811c6d3bff69@thread.skype_p_FjjyhkqXg0y7gb3USpRpM2UADlkA_h_1615327590963?tenantId=9093f1a3-8771-4fb7-8596-d51eeef18cda&webUrl=https%3A%2F%2Ftasks.teams.microsoft.com%2Fteamsui%2FpersonalApp%2Falltasklists&context=%7B%22subEntityId%22%3A%22%2Fboard%2Ftask%2FDGeOxivaGEmgRVf0YMx7mmUACpnT%22%2C%22channelId%22%3A%2219%3A2564d72f7a7e4ab2b45d811c6d3bff69%40thread.skype%22%7D

**Pull request merged**